service/tablet_allocator.cc

```diff
@@ -1049,10 +1083,40 @@ class load_balancer {
                     return r2_hosts.contains(dst.host);
                 };
 
-                lblogger.debug("Replica sets of tablets being co-located: ({}: {}), ({}, {})",
-                        global_tablet_id{table, t1.tid}, r1, global_tablet_id{table, t2.tid}, r2);
+                bool rack_awareness = rack_aware_colocation();
 
-                maybe_generate_colocation_plan({t1, std::move(r1)}, {t2, std::move(r2)}, check_constraints);
+                lblogger.debug("Replica sets of tablets being co-located: ({}: {}), ({}, {}), rack_awareness={}",
+                        global_tablet_id{table, t1.tid}, r1, global_tablet_id{table, t2.tid}, r2, rack_awareness);
+
+                if (rack_awareness) {
+                    auto r1_sets = per_rack_replica_sets(r1);
+                    auto r2_sets = per_rack_replica_sets(r2);
+
+                    for (auto&& [rack, r1_s] : r1_sets) {
+                        auto it = r2_sets.find(rack);
+                        if (it == r2_sets.end()) {
+                            // When adding a new rack, e.g. from RF-1 to RF, only one of sibling tables might have
+                            // presence on all racks, so we can wait for migrations to settle before we proceed
+                            // with their co-location.
+                            lblogger.warn("Unable to find replicas of both sibling tablets on rack {}: ({}: {}), ({}, {})",
+                                    rack, global_tablet_id{table, t1.tid}, r1, global_tablet_id{table, t2.tid}, r2);
+                            return make_ready_future<>();
+                        }
+                        auto& r2_s = it->second;
+                        // Skip to next rack if current one is co-located.
+                        if (r1_s == r2_s) {
+                            continue;
+                        }
+
+                        // Each tablet allows for only one transition, so continue to next set of sibling tablets
+                        // if a co-location plan was generated for current one.
```

Comment:
I'm confused by this comment. It seems like we should skip to the next rack if a plan was **not** generated for current one, isn't it?

Reply:
That's also true. What the commend says is that if a co-location migration was generated for a given pair of sibling tablets, then there's no need to look for further migrations for that same pair, we can move to the next one. I will reword the comment.

Reply:
ah, to the next set of sibling tablets, not continue in the loop...
it's a bit confusing.